lib/auth/oidc_test.go

```diff
@@ -186,143 +186,174 @@ func TestUserInfoBadStatus(t *testing.T) {
+			values := url.Values{
+				"code":  []string{"XXX-code"},
+				"state": []string{request.StateToken},
+			}
 
-	diagCtx := ssoDiagContext{}
+			// override getClaimsFun.
+			s.a.getClaimsFun = func(closeCtx context.Context, oidcClient *oidc.Client, connector types.OIDCConnector, code string) (jose.Claims, error) {
+				cc := map[string]interface{}{
+					"email_verified": true,
+					"groups":         []string{"everyone", "idp-admin", "idp-dev"},
+					"email":          "superuser@example.com",
+					"sub":            "00001234abcd",
+					"exp":            1652091713.0,
+				}
+				return cc, nil
+			}
 
-	resp, err = s.a.validateOIDCAuthCallback(ctx, &diagCtx, values)
-	require.NoError(t, err)
-	require.NotNil(t, resp)
-	require.Equal(t, &OIDCAuthResponse{
-		Username: "superuser@example.com",
-		Identity: types.ExternalIdentity{
-			ConnectorID: "-sso-test-okta",
-			Username:    "superuser@example.com",
-		},
-		Req: *request,
-	}, resp)
-	require.Equal(t, types.SSODiagnosticInfo{
-		TestFlow: true,
-		Success:  true,
-		CreateUserParams: &types.CreateUserParams{
-			ConnectorName: "-sso-test-okta",
-			Username:      "superuser@example.com",
-			Logins:        nil,
-			KubeGroups:    nil,
-			KubeUsers:     nil,
-			Roles:         []string{"access"},
-			Traits: map[string][]string{
-				"email":  {"superuser@example.com"},
-				"groups": {"everyone", "idp-admin", "idp-dev"},
-				"sub":    {"00001234abcd"},
-			},
-			SessionTTL: 600000000000,
-		},
-		OIDCClaimsToRoles: []types.ClaimMapping{
-			{
-				Claim: "groups",
-				Value: "idp-admin",
-				Roles: []string{"access"},
-			},
-		},
-		OIDCClaimsToRolesWarnings: nil,
-		OIDCClaims: map[string]interface{}{
-			"email_verified": true,
-			"groups":         []string{"everyone", "idp-admin", "idp-dev"},
-			"email":          "superuser@example.com",
-			"sub":            "00001234abcd",
-			"exp":            1652091713.0,
-		},
-		OIDCIdentity: &types.OIDCIdentity{
-			ID:        "00001234abcd",
-			Name:      "",
-			Email:     "superuser@example.com",
-			ExpiresAt: diagCtx.info.OIDCIdentity.ExpiresAt,
-		},
-		OIDCTraitsFromClaims: map[string][]string{
-			"email":  {"superuser@example.com"},
-			"groups": {"everyone", "idp-admin", "idp-dev"},
-			"sub":    {"00001234abcd"},
-		},
-		OIDCConnectorTraitMapping: []types.TraitMapping{
-			{
-				Trait: "groups",
-				Value: "idp-admin",
-				Roles: []string{"access"},
-			},
-		},
-	}, diagCtx.info)
+			resp, err := s.a.ValidateOIDCAuthCallback(ctx, values)
+			if tc.err != nil {
+				require.ErrorIs(t, err, tc.err)
+				return
+			}
+
+			require.NoError(t, err)
+			require.NotNil(t, resp)
+			require.Equal(t, &OIDCAuthResponse{
+				Username: "superuser@example.com",
+				Identity: types.ExternalIdentity{
+					ConnectorID: "-sso-test-okta",
+					Username:    "superuser@example.com",
+				},
+				Req: *request,
+			}, resp)
+
+			diagCtx := ssoDiagContext{}
+
+			resp, err = s.a.validateOIDCAuthCallback(ctx, &diagCtx, values)
```

Comment:
Isn't this called as part of the public function above? Why do we go through it again?

Reply:
Not sure. The double validate was already part of the test before this PR.